src/main/java/baritone/process/BuilderProcess.java

```diff
@@ -113,21 +113,24 @@ public void build(String name, ISchematic schematic, Vec3i origin) {
         this.origin = new Vec3i(x, y, z);
         this.paused = false;
         this.layer = Baritone.settings().startAtLayer.value;
+        this.stopAtHeight = schematic.heightY();
         if (Baritone.settings().buildOnlySelection.value) {
             if (baritone.getSelectionManager().getSelections().length == 0) {
                 logDirect("Poor little kitten forgot to set a selection while BuildOnlySelection is true");
-            }
-            if (Baritone.settings().buildInLayers.value) {
+                this.stopAtHeight = 0;
+            } else if (Baritone.settings().buildInLayers.value) {
                 OptionalInt minim = Stream.of(baritone.getSelectionManager().getSelections()).mapToInt(sel -> sel.min().y).min();
                 OptionalInt maxim = Stream.of(baritone.getSelectionManager().getSelections()).mapToInt(sel -> sel.max().y).max();
-                logDirect(String.format("Schematic starts at y=%s with height %s", y, schematic.heightY()));
-                logDirect(String.format("Selection starts at y=%s and ends at y=%s", minim.isPresent() ? minim.getAsInt() : 0, maxim.isPresent() ? maxim.getAsInt() : 0));
                 if (minim.isPresent() && maxim.isPresent()) {
-                    selectionYBounds[0] = Baritone.settings().layerOrder.value ? y + schematic.heightY() - maxim.getAsInt() : minim.getAsInt() - y;
-                    selectionYBounds[1] = Baritone.settings().layerOrder.value ? y + schematic.heightY() - minim.getAsInt() : maxim.getAsInt() - y;
+                    int startAtHeight = Baritone.settings().layerOrder.value ? y + schematic.heightY() - maxim.getAsInt() : minim.getAsInt() - y;
+                    this.stopAtHeight = (Baritone.settings().layerOrder.value ? y + schematic.heightY() - minim.getAsInt() : maxim.getAsInt() - y) + 1;
+                    this.layer = startAtHeight / Baritone.settings().layerHeight.value;
+                    if (Baritone.settings().chatDebug.value) {
+                        logDirect(String.format("Schematic starts at y=%s with height %s", y, schematic.heightY()));
+                        logDirect(String.format("Selection starts at y=%s and ends at y=%s", minim.getAsInt(), maxim.getAsInt()));
+                        logDirect(String.format("Considering relevant height %s - %s", startAtHeight, this.stopAtHeight));
+                    }
```

Comment:
There's also `logDebug`, which does the same without manually checking `chatDebug`.